react/features/virtual-background/components/VirtualBackgroundDialog.js

```diff
@@ -54,23 +55,73 @@ type Props = {
  * @returns {ReactElement}
  */
 function VirtualBackground({ dispatch, t }: Props) {
+    const localImages = localStorage.getItem('storedImages');
+    const [ storedImages, setStoredImages ] = useState((localImages && JSON.parse(localImages)) || []);
+    const [ loading, isloading ] = useState(false);
+
+
+    /**
+     * Updates stored images on local storage.
+     */
+    useEffect(() => {
+        localStorage.setItem('storedImages', JSON.stringify(storedImages));
+    }, [ storedImages ]);
+
     const [ selected, setSelected ] = useState('');
-    const enableBlur = () => {
+    const enableBlur = async () => {
+        isloading(true);
         setSelected('blur');
-        dispatch(setVirtualBackground('', false));
-        dispatch(toggleBackgroundEffect(true));
+        await dispatch(setVirtualBackground('', false));
+        await dispatch(toggleBackgroundEffect(true));
+        isloading(false);
     };
 
-    const removeBackground = () => {
+    const removeBackground = async () => {
+        isloading(true);
         setSelected('none');
-        dispatch(setVirtualBackground('', false));
-        dispatch(toggleBackgroundEffect(false));
+        await dispatch(setVirtualBackground('', false));
+        await dispatch(toggleBackgroundEffect(false));
+        isloading(false);
     };
 
-    const addImageBackground = image => {
+    const addImageBackground = async image => {
+        isloading(true);
         setSelected(image.id);
-        dispatch(setVirtualBackground(image.src, true));
-        dispatch(toggleBackgroundEffect(true));
+        await dispatch(setVirtualBackground(image.src, true));
+        await dispatch(toggleBackgroundEffect(true));
+        isloading(false);
+    };
+
+    const deleteStoredImage = image => {
+        setStoredImages(storedImages.filter(item => item !== image));
+    };
+
+    const uploadImage = async imageFile => {
+        const reader = new FileReader();
+
+        reader.readAsDataURL(imageFile[0]);
+        reader.onload = async () => {
+            isloading(true);
+            setStoredImages([
+                ...storedImages,
+                {
+                    tooltip: imageFile[0].name,
+                    name: imageFile[0].name,
+                    id: Math.random()
+                        .toString(36)
+                        .substring(2, 7),
+                    src: reader.result.toString()
```

Comment:
I put toString() function because the virtualSource from action file is set to be a string and the base64(from uploaded image) is ArrayBuffer. I can remove this toString() function and let the type of virtualSource be 'any'

Reply:
We could use either, but I see we use it in 2 places, ideally we only do this conversion once.